infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/provision/AsyncStoragePodWatcher.java

```diff
@@ -0,0 +1,171 @@
+ * which is available at https://www.eclipse.org/legal/epl-2.0/
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *
+ * Contributors:
+ *   Red Hat, Inc. - initial API and implementation
+ */
+
+package org.eclipse.che.workspace.infrastructure.kubernetes.provision;
+
+import static com.google.common.base.Strings.isNullOrEmpty;
+import static java.lang.Long.parseLong;
+import static org.eclipse.che.api.core.Pages.iterateLazily;
+import static org.eclipse.che.api.workspace.shared.Constants.LAST_ACTIVE_INFRASTRUCTURE_NAMESPACE;
+import static org.eclipse.che.api.workspace.shared.Constants.LAST_ACTIVITY_TIME;
+import static org.eclipse.che.workspace.infrastructure.kubernetes.namespace.pvc.CommonPVCStrategy.COMMON_STRATEGY;
+import static org.eclipse.che.workspace.infrastructure.kubernetes.provision.AsyncStorageProvisioner.ASYNC_STORAGE;
+
+import io.fabric8.kubernetes.api.model.DoneablePod;
+import io.fabric8.kubernetes.api.model.Pod;
+import io.fabric8.kubernetes.client.dsl.PodResource;
+import java.time.Clock;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.TimeUnit;
+import javax.inject.Inject;
+import javax.inject.Named;
+import javax.inject.Singleton;
+import org.eclipse.che.api.core.ServerException;
+import org.eclipse.che.api.core.model.user.User;
+import org.eclipse.che.api.user.server.PreferenceManager;
+import org.eclipse.che.api.user.server.UserManager;
+import org.eclipse.che.api.workspace.server.WorkspaceRuntimes;
+import org.eclipse.che.api.workspace.server.spi.InfrastructureException;
+import org.eclipse.che.api.workspace.server.spi.InternalRuntime;
+import org.eclipse.che.commons.annotation.Nullable;
+import org.eclipse.che.commons.schedule.ScheduleDelay;
+import org.eclipse.che.workspace.infrastructure.kubernetes.KubernetesClientFactory;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Periodically checks ability to stop Asynchronous Storage Pod. It will periodically revise
+ * UserPreferences of all registered user and check specialized preferences. Preferences should be
+ * recorded if last workspace stopped and cleanup on start any workspace. Required preferences to
+ * initiate stop procedure for Asynchronous Storage Pod : {@link
+ * org.eclipse.che.api.workspace.shared.Constants#LAST_ACTIVE_INFRASTRUCTURE_NAMESPACE} : should
+ * contain last used infrastructure namespace {@link
+ * org.eclipse.che.api.workspace.shared.Constants#LAST_ACTIVITY_TIME} : seconds then workspace
+ * stopped in the Java epoch time format (aka Unix time)
+ */
+@Singleton
+public class AsyncStoragePodWatcher {
+
+  private static final Logger LOG = LoggerFactory.getLogger(AsyncStoragePodWatcher.class);
+
+  private final KubernetesClientFactory kubernetesClientFactory;
+  private final UserManager userManager;
+  private final PreferenceManager preferenceManager;
+  private final WorkspaceRuntimes runtimes;
+  private final long shutdownTimeoutSec;
+  private final Clock clock = Clock.systemDefaultZone();
```

Comment:
What's a benefit of having Clock here? It's meant to use for easier testing, but you're not using it for that.

Reply:
Yes, maybe later some tests will be added

Reply:
for that to be useful, it would have to be injected so you can put here some custom Clock instance for testing purpose. But you're using it only to get instant at one place. What's the benefit over calling `Instant.now()` at that place ?

Reply:
yep got it